<launch>

    <!-- Launch Robot_State_Publisher -->
    <include file="$(find my_bot)/launch/my_bot.launch"/>

    <arg name="debug" default="false" />
    <arg name="gui" default="true" />
    <arg name="pause" default="false" />
    <arg name="world" default="$(find my_bot)/worlds/newArena.world" />
    <arg name="params_file" default="/home/william/catkin_ws/src/my_bot/config/mapper_params_online_async.yaml"/>

    <!-- Launch Gazebo -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="world_name" value="$(arg world)" />
        <arg name="debug" value="$(arg debug)" />
        <arg name="gui" value="$(arg gui)" />
        <arg name="paused" value="$(arg pause)" />
        <arg name="use_sim_time" value="true" />
    </include>    

    <!-- Spawn Robot -->
    <arg name="x" default="1"/>
    <arg name="y" default="1"/>
    <arg name="z" default="0"/>

    <node name="mybot_spawn" pkg="gazebo_ros" type="spawn_model" output="screen"
        args="-urdf -param robot_description -model my_bot -x $(arg x) -y $(arg y) -z $(arg z)"/>
 
    <!-- Distance Sensor Wrapper -->
    <node name="distance_sens_wrapper" pkg="my_bot" type="distance_sens_wrapper" output="screen" >
        <param name="left_destination_frame" value="distance_sensor_left"/>
        <param name="right_destination_frame" value="distance_sensor_right"/>
        <param name="left_destination_topic" value="/new_left_scan"/>
        <param name="right_destination_topic" value="/new_right_scan"/>
        <param name="angle_min" value="-0.139626"/>
        <param name="angle_max" value="0.139626"/>
        <param name="angle_increment" value="0.0116355"/>
        <param name="scan_time" value="0.1"/>
        <param name="range_min" value="0.15"/>
        <param name="range_max" value="10.0"/>
        <param name="detection_threshold" value="2.1"/>
    </node>

    <!-- Laser Merger -->
    <node pkg="ira_laser_tools" name="laserscan_multi_merger" type="laserscan_multi_merger" output="screen">
        <param name="destination_frame" value="laser_frame"/>
        <param name="cloud_destination_topic" value="/merged_cloud"/>
        <param name="scan_destination_topic" value="/scan_multi"/>
        <param name="laserscan_topics" value ="/lidar_scan /new_left_scan /new_right_scan" /> <!-- LIST OF THE LASER SCAN TOPICS TO SUBSCRIBE -->
        <param name="angle_min" value="-1.71"/>
        <param name="angle_max" value="1.71"/>
        <param name="angle_increment" value="0.008159980918351739"/>
        <param name="scan_time" value="0.1"/>
        <param name="range_min" value="0.10"/>
        <param name="range_max" value="10.0"/>
	</node>

    <!-- SLAM -->
    <node pkg="slam_toolbox" type="sync_slam_toolbox_node" name="slam_toolbox" output="screen">
      <rosparam command="load" file="$(arg params_file)" />
    </node>    

    <!-- MOVE_BASE -->
    <include file="$(find mars_2dnav)/move_base.launch"/>

</launch>